File: sfs/public/js/sales_invoice.js
frappe.ui.form.on("Sales Invoice", {
    onload_post_render:function(frm){
        console.log("onload")
        if (frm.doc.new_doc==0){
            if(cur_frm.doc.timesy_list && cur_frm.doc.company){
                frappe.call({
                    method: "sfs.doc_events.sales_invoice.get_timesy_dates",
                    args: {
                        name:cur_frm.doc.timesy_list[0].timesy,
                        company:cur_frm.doc.company
                    },
                    callback: function (r) {
                        if(! cur_frm.doc.s_date){
                            cur_frm.set_value("s_date",r.message[0].start_date)
                            
                        }
                        console.log("onload woking")
                        cur_frm.set_value("additional_amount",r.message[0].charge_amount)
                        if(! cur_frm.doc.e_date){
                            cur_frm.set_value("e_date",r.message[0].end_date)
                        }
                        if (cur_frm.doc.items && cur_frm.doc.hourly_invoice==0){
                            var items = cur_frm.doc.items
                            for (var i=0;i<items.length;i++){
                                items[i].reference_type=r.message[0].reference_type
                                items[i].employee=r.message[0].employee_code

                                items[i].employee_name=r.message[0].employee_name
                                items[i].staff=r.message[0].staff_code
                                items[i].custom_timesy = r.message[0].name
                                items[i].staff_name=r.message[0].staff_name

                                items[i].iqama_id=r.message[0].iq_id,
                                items[i].nationality=r.message[0].nation,
                                items[i].staff_iqama_id=r.message[0].s_iq_id,
                                items[i].staff_nationality=r.message[0].s_nation,
                                items[i].item_name=r.message[0].item_name,
                                items[i].uom=r.message[0].uom,
                                items[i].qty=1,
                                items[i].rate=r.message[0].price_list_rate,
                                items[i].timesy_rate=r.message[0].total_costing_rate_before_deduction,
                                items[i].amount=r.message[0].price_list_rate*1,
                                items[i].total_working_hour=r.message[0].total_working_hour,
                                items[i].conversion_factor=1,
                                items[i].description=r.message[0].description,
                                items[i].income_account = r.message[0].income_account,
                                items[i].hourly_rate = r.message[0].hourly_rate


                            }
                        }
                        if (cur_frm.doc.items && cur_frm.doc.hourly_invoice==1){
                            var items = cur_frm.doc.items
                            for (var i=0;i<items.length;i++){
                                items[i].reference_type=r.message[0].reference_type
                                items[i].employee=r.message[0].employee_code

                                items[i].employee_name=r.message[0].employee_name
                                items[i].staff=r.message[0].staff_code

                                items[i].staff_name=r.message[0].staff_name
                                items[i].custom_timesy = r.message[0].name
                                items[i].iqama_id=r.message[0].iq_id,
                                items[i].nationality=r.message[0].nation,
                                items[i].staff_iqama_id=r.message[0].s_iq_id,
                                items[i].staff_nationality=r.message[0].s_nation,
                                items[i].item_name=r.message[0].item_name,
                                items[i].uom=r.message[0].uom,
                                items[i].qty=1,
                                items[i].hourly_rate = r.message[0].hourly_rate,
                                items[i].rate=r.message[0].total_costing_rate_before_deduction,
                                items[i].timesy_rate=r.message[0].total_costing_rate_before_deduction,
                                items[i].amount=r.message[0].total_costing_rate_before_deduction * items[i].qty,
                                items[i].total_working_hour=r.message[0].total_working_hour,
                                items[i].conversion_factor=1,
                                items[i].description=r.message[0].description,
                                items[i].income_account = r.message[0].income_account
                               


                            }
                        }
                        
                    }
                })
            }
            
        }
        cur_frm.add_custom_button(__('Timesy'),
        function() {
            var query_args = {
                query:"sfs.doc_events.sales_invoice.get_staffing",
                filters: {
        doctype: cur_frm.doc.doctype
                }
            }
                var d = new frappe.ui.form.MultiSelectDialog({
                        doctype: "Timesy",
                        target: cur_frm,
                        setters: {
                            staffing_type: "",
                            customer_name: null,
                            employee_name: null,
                            start_date: null,
                        },
                        date_field: "start_date",
                        get_query() {
                            return query_args;
                        },
                        action(selections) {
                            console.log(selections)
                            console.log("sfs executed")
                            add_timesy(selections, cur_frm)
                            console.log("dates")
                            
                            // add_dates(selections,cur_frm)
                            // console.log("items")
                            // get_items(selections,cur_frm)
                            d.dialog.hide()
                        }
                    });
    }, __("Get Items From"), "btn-default");

    },
    refresh: function () {
        console.log('done')
        frappe.call({
            method: "sfs.doc_events.sales_invoice.get_timesies",
            args: {doctype: "Sales Invoice"},
            callback: function (r) {
                cur_frm.set_query("timesy","timesy_list", () => {
                    return {
                        filters: [
                            ["name", "not in", r.message],
                            ["status", "=", "Completed"],
                            ["docstatus", "=", 1],
                        ]
                    }
                })
            }
        })


        
        cur_frm.add_custom_button(__('Fetch from Timesy'),
				function() {
                    console.log("hiiiii ranju")
                    console.log(cur_frm.doc.customer_name)
                    var query_args = {
                       query:"sfs.doc_events.sales_invoice.get_staffing",
                        // filters: {doctype: cur_frm.doc.doctype}
                    }
					 var d = new frappe.ui.form.MultiSelectDialog({
                                doctype: "Timesy",
                                target: cur_frm,
                                setters: {
                                    staffing_type: "",
                                    employee_name: null,
                                    customer_name: cur_frm.doc.customer_name,
                                    start_date: null,
                                },
                                date_field: "start_date",
                                get_query() {
                                    return query_args;
                                },
                                action(selections) {
                                    add_items(selections, cur_frm)
                                    d.dialog.hide()
                                }
                            });
        }, __("Get Items From"), "btn-default");



        cur_frm.add_custom_button(__('Timesy'),
                function() {
                    var query_args = {
                        query:"sfs.doc_events.sales_invoice.get_staffing",
                        filters: {
                doctype: cur_frm.doc.doctype,
                        }
                    }
                        var d = new frappe.ui.form.MultiSelectDialog({
                                doctype: "Timesy",
                                target: cur_frm,
                                setters: {
                                    staffing_type: "",
                                    customer_name: null,
                                    employee_name: null,
                                    start_date: null,
                                },
                                date_field: "start_date",
                                get_query() {
                                    return query_args;
                                },
                                action(selections) {
                                    console.log(selections)
                                    console.log("sfs executed")
                                    add_timesy(selections, cur_frm)
                                    console.log("dates")
                                    
                                    add_dates(selections,cur_frm)
                                    console.log("items")
                                    get_items(selections,cur_frm)
                                    d.dialog.hide()
                                }
                            });
        }, __("Get Items From"), "btn-default");
    },
    hourly_invoice:function(frm){
        if(cur_frm.doc.hourly_invoice==1){
            var items = cur_frm.doc.items
            for(var i=0 ; i<items.length ; i++){
                let r = items[i].hourly_rate*items[i].total_working_hour
                items[i].rate = r
                items[i].amount = r * items[i].qty

            }
            cur_frm.refresh_field("items")
        }
        if(cur_frm.doc.hourly_invoice==0){
            var items = cur_frm.doc.items
            for(var i=0 ; i<items.length ; i++){
                let r =0
                r = items[i].price_list_rate
                items[i].rate = r
                if (r){
                items[i].amount = r * items[i].qty
                }
                else{
                    items[i].amount =0

                }

            }
            cur_frm.refresh_field("items")

        }
    },
    validate:function(frm){
        frm.set_value("new_doc",1)
    }
    
})

frappe.ui.form.on("Sales Invoice Item", {
    total_working_hour:function(frm,cdt,cdn){
        var d = locals[cdt][cdn]
        if(cur_frm.doc.hourly_invoice==1){
            if (d.hourly_rate){
                d.rate = d.total_working_hour * d.hourly_rate
                d.amount = (d.hourly_rate * d.total_working_hour)*d.qty
            }
            cur_frm.refresh_field("items")
            total_cal(cur_frm)

        }
    },
    hourly_rate:function(frm,cdt,cdn){
        var d = locals[cdt][cdn]
        if(cur_frm.doc.hourly_invoice==1){
            if (d.total_working_hour){
                d.rate = d.hourly_rate * d.total_working_hour
                d.amount = (d.hourly_rate * d.total_working_hour)*d.qty
            }
            cur_frm.refresh_field("items")
            total_cal(cur_frm)

        }
    },

})
frappe.ui.form.on("Timesy List", {
    timesy_list_remove: function () {
        compute_grand_costing(cur_frm)
    },
    timesy: function (frm, cdt, cdn) {

        var d = locals[cdt][cdn]
        if(d.timesy){
        frappe.db.get_doc("Timesy",d.timesy)
            .then(t => {
                d.staff_name = t.reference_type === 'Staff' ? t.staff_name : t.employee_name
                cur_frm.refresh_field(d.parentfield)
            compute_grand_costing(cur_frm)
        })
        }

    }
})
    
function total_cal(cur_frm){
    var docs = cur_frm.doc.items
    var total = 0
    for (var i=0;i<docs.length;i++){
        total += docs[i].amount 
    }
    cur_frm.set_value("total",total)
} 
    
function add_timesy(selections, cur_frm) {
        if(!cur_frm.doc.company){
            frappe.throw("Please select the Company first")
        }
        frappe.call({
            method: "sfs.doc_events.sales_invoice.get_timesy",
            args: {
                name: selections
            },
            callback: function (r) {
                console.log("yess.................")
                for(var x=0;x<r.message.length;x+=1){
                    cur_frm.add_child("timesy_list", {
                        timesy: r.message[x].name,
                        staff_name: r.message[x].staff_name,
                        staffing_project: r.message[x].staffing_project,
                        total_costing_rate: r.message[x].total_costing_hour,
                    })
                    cur_frm.refresh_field("timesy_list")
                    
                    
                    compute_grand_costing(cur_frm)
                }
                var d = 0
                for(var x=0;x<r.message.length;x+=1){
                    d += r.message[x].total_deduction
                }
                var a = 0
                for(var s=0;s<r.message.length;s+=1){
                    a += r.message[s].charge_amount
                }
                cur_frm.set_value("discount_amount",d)
                console.log("aaaaaa")
                console.log(a)
                cur_frm.set_value("additional_amount",a)

                add_dates(selections,cur_frm)
                console.log("items")
                get_items(selections,cur_frm)
            }
        })
}
function check_items(item, cur_frm) {
        for(var x=0;x<cur_frm.doc.items.length;x+=1){
            var item_row = cur_frm.doc.items[x]
            if(item_row.item_code === item){
                item_row.qty += 1
                cur_frm.refresh_field("items")
                return true
            }
        }
        return false
}
function compute_grand_costing(cur_frm) {
    var total = 0
    for(var x=0;x<cur_frm.doc.timesy_list.length;x+=1){
        total += cur_frm.doc.timesy_list[x].total_costing_rate
    }
    cur_frm.doc.grand_costing_rate = total
    cur_frm.refresh_field("grand_costing_rate")
    if(cur_frm.doc.items.length > 0){
        cur_frm.doc.items[cur_frm.doc.items.length - 1].rate = total
        cur_frm.refresh_field('items')
    }
}
function add_dates(selections, cur_frm) {

    frappe.call({
        method: "sfs.doc_events.sales_invoice.get_dates",
        args: {
            name: selections
        },
        callback: function (r) {
            console.log("rrrrrrrrrrrr")
            console.log(r.message)
            cur_frm.set_value("s_date",r.message[0].start_date)
            cur_frm.set_value("e_date",r.message[0].end_date)

        }
    })
}


function get_items(selections, cur_frm) {
    console.log("get items")
    frappe.call({
        method: "sfs.doc_events.sales_invoice.get_item_details",
        args: {
            name: selections,
            company:cur_frm.doc.company
        },
        callback: function (r) {
            console.log("hiiii")
            console.log(r.message)
            for(var i=0;i<cur_frm.doc.items.length;i+=1){
                if(!cur_frm.doc.items[i].item_code){
                    cur_frm.clear_table("items")
                }
            }
            cur_frm.set_value("hourly_invoice",1)
            for(var x=0;x<r.message.length;x+=1){
                cur_frm.add_child("items", {
                    item_code: r.message[x].item,
                    description:r.message[x].description,
                    reference_type:r.message[x].reference_type,
                    custom_timesy:r.message[x].name,
                    staff:r.message[x].staff_code,
                    staff_name:r.message[x].staff_name,
                    employee:r.message[x].employee_code,
                    employee_name:r.message[x].employee_name,
                    iqama_id:r.message[x].iq_id,
                    nationality:r.message[x].nation,
                    staff_iqama_id:r.message[x].s_iq_id,
                    staff_nationality:r.message[x].s_nation,
                    item_name:r.message[x].item_name,
                    uom:r.message[x].uom,
                    qty:1,
                    rate:r.message[x].total_working_hour*r.message[x].price_list_rate,
                    timesy_rate:r.message[x].total_costing_hour,
                    amount:r.message[x].total_working_hour*r.message[x].price_list_rate*1,
                    total_working_hour:r.message[x].total_working_hour,
                    conversion_factor:1,
                    income_account:r.message[x].income_account,
                    hourly_rate:r.message[x].hourly_rate
                })
                cur_frm.refresh_field("items")
                // compute_grand_costing(cur_frm)
            }
        }
    })
}



function add_items(selections, cur_frm) {
    console.log("working add items")
    frappe.call({
            method: "sfs.doc_events.sales_invoice.get_bulk_timesy",
            args: {
                name: selections
            },
            callback: function (r) {
                if (!cur_frm.doc.items[0].item_code){
                    cur_frm.clear_table("items")
                }
                for(var x=0;x<r.message.length;x+=1){
                    console.log("super 999999999")
                    console.log(r.message[x].description)
                    cur_frm.add_child("items", {
                        "item_code":r.message[x].item,
                        "item_name":r.message[x].item_name,
                        "description":r.message[x].description,
                        "income_account":r.message[x].income_account,
                        "uom":r.message[x].stock_uom,
                        "custom_timesy":r.message[x].name,
                        "reference_type":r.message[x].reference_type,
                        "employee":r.message[x].employee_code,
                        "staff":r.message[x].staff_code,           
                        "employee_name":r.message[x].employee_name,
                        "staff_name":r.message[x].staff_name,
                        "iqama_id":r.message[x].iqama_id,
                        "staff_iqama_id":r.message[x].staff_iqama_id,
                        "nationality":r.message[x].nationality,
                        "staff_nationality":r.message[x].staff_nationality,
                        "total_working_hour":r.message[x].total_working_hour,
                        "hourly_rate":r.message[x].hourly_rate,
                        "qty":1,
                        "rate":r.message[x].total_working_hour * r.message[x].hourly_rate,
                        "amount":r.message[x].total_working_hour * r.message[x].hourly_rate,
                        "timesy_rate":r.message[x].total_costing_rate_before_deduction,
                        "type":"Regular"
                         })
                    cur_frm.refresh_field("items")
                    total_calculation(cur_frm);
                   
                }
            }
        })
}



function total_calculation(cur_frm){
    var table = cur_frm.doc.items
    var total = 0
    var t_qty = 0
    for (var i =0 ;i<table.length;i++){
        if(table[i].amount){
            total += table[i].amount
            t_qty += table[i].qty
            
        }
    }
    cur_frm.set_value("total_qty",t_qty)
    cur_frm.set_value("total",total)
    if(cur_frm.doc.total_taxes_and_charges){
        cur_frm.set_value("grand_total",total+cur_frm.doc.total_taxes_and_charges)
        cur_frm.set_value("rounded_total",total+cur_frm.doc.total_taxes_and_charges)
    }
    else{
        cur_frm.set_value("grand_total",total)
        cur_frm.set_value("rounded_total",total)
    }
    


}

File: sfs/doc_events/sales_invoice.py
import frappe, json
@frappe.whitelist()
def get_timesies(doctype):
    query = """ SELECT T.name FROM `tabTimesy` T 
                          INNER JOIN `tabTimesy List` TL ON TL.timesy = T.name 
                          INNER JOIN `tab{0}` DD ON DD.name = TL.parent
                          WHERE T.docstatus = 1 and DD.docstatus = 1 and T.status = 'Completed'""".format(doctype)
    t = frappe.db.sql(query, as_dict=1)

    return [i.name for i in t]
def get_condition(filters):
    if filters['staffing_type'][1]:
        if filters['staffing_type'][0] == "in" or filters['staffing_type'][0] == "not in":
            in_not_in = ()
            if len(filters['staffing_type'][1]) == 1:
                return " and staffing_type = '{0}' ".format(filters['staffing_type'][1][0])

            for i in filters['staffing_type'][1]:
                in_not_in += (i,)
            return " and staffing_type {0} {1} ".format(filters['staffing_type'][0],in_not_in)
        return " and staffing_type {0} '{1}' ".format(filters['staffing_type'][0],filters['staffing_type'][1])
    return ""
@frappe.whitelist()
def get_staffing(doctype, target,setters,d,e,filters):
    print("============================")
    print(doctype)
    print(target)
    print(setters)
    print(d)
    print(e)
    print(filters)
    data = []
    condition = "" 
    if target:
        condition += " and (name like '%{0}%') ".format(target)

    if 'customer_name' in filters:
        condition += " and customer_name like '{0}' ".format(filters['customer_name'][1])

    if "staffing_type" in filters:
        condition += get_condition(filters)

    if "start_date" in filters:
        condition += " and start_date = '{0}' ".format(filters['start_date'])

    if "supplier_name" in filters:
        condition += " and supplier_name like '{0}' ".format(filters['supplier_name'][1])

    if "employee_name" in filters:
        condition += " and (employee_name like '{0}' or staff_name like '{1}') ".format(filters['employee_name'][1],filters['employee_name'][1])

    time_list = ""
    if "doctype" in filters:
        time_list += " and parenttype = '{0}'".format(filters['doctype'])

    print(condition)
    query = """ SELECT * FROM `tabTimesy` WHERE docstatus=1 and status='Completed' {0}""".format(condition)
    print(query)
    staffing = frappe.db.sql(query, as_dict=1)
    for i in staffing:
        check = frappe.db.sql(""" SELECT * FROm `tabTimesy List` WHERE timesy=%s {0}""".format(time_list), i.name, as_dict=1)
        if len(check) == 0:
            data.append({
                "name": i.name,
                "staffing_type": i.staffing_type,
                "employee_name": i.staff_name if i.reference_type == 'Staff' else i.employee_name,
                "customer_name": i.customer_name,
                "supplier_name": i.supplier_name,
                "start_date": i.start_date
            })
    print(data)
    return data


@frappe.whitelist()
def get_timesy(name):
    data = json.loads(name)
    timesies = []

    for i in data:
        timesies += frappe.db.sql(""" SELECT * FROM `tabTimesy` WHERE name=%s""", i, as_dict=1)
    ded = 0


    return timesies

@frappe.whitelist()
def get_price(item):
    price = frappe.db.sql("""select price_list_rate from `tabItem Price` where item_code=%s order by valid_from desc limit 1""",item,as_dict=1)
    if price:
        item_price = price[0].price_list_rate
    return item_price


@frappe.whitelist()
def get_dates(name):
    data = json.loads(name)
    dates = []

    for i in data:
        dates = frappe.db.sql(""" SELECT start_date,end_date FROM `tabTimesy` WHERE name=%s""", i, as_dict=1)
    return dates

@frappe.whitelist()
def get_timesy_dates(name,company):
    print("*****************")
    datas = frappe.db.sql(""" SELECT name,item,total_working_hour,total_deduction,total_costing_rate_before_deduction,total_costing_hour,start_date,end_date,employee_code,
    employee_name,staff_code,staff_name,reference_type,charge_amount FROM `tabTimesy` WHERE name=%s""", name, as_dict=1)
    for i in datas:
        s_iq_id=''
        s_nat = ''
        iq_id=''
        nat = ''
        if i.employee_code:    
            iq_id = frappe.db.get_value("Employee",i.employee_code,"iqama_id") 
            nation = frappe.db.get_value("Employee",i.employee_code,"nationality") 
            i_id ={"iq_id":iq_id}
            i.update(i_id)
            nat ={"nation":nation}
            i.update(nat)
            price = frappe.db.sql("""select default_cost_rate_per_hour from `tabStaffing Cost` where employee_code=%s and status='Active'""",i.employee_code,as_dict=1)
            if price:
                rate = {"hourly_rate":price[0].default_cost_rate_per_hour}
                i.update(rate)
            else:
                rate = {"hourly_rate":0}
                i.update(rate)
        if i.staff_code:
            
            s_iq_id = frappe.db.get_value("Staff",i.staff_code,"iqama_number") 
            s_nation = frappe.db.get_value("Staff",i.staff_code,"nationality") 
            s_i_id ={"s_iq_id":s_iq_id}
            i.update(s_i_id)
            s_nat ={"s_nation":s_nation}
            i.update(s_nat)
            price = frappe.db.sql("""select default_cost_rate_per_hour from `tabStaffing Cost` where staff_code=%s and status='Active'""",i.staff_code,as_dict=1)
            if price:
                rate = {"hourly_rate":price[0].default_cost_rate_per_hour}
                i.update(rate)
            else:
                rate = {"hourly_rate":0}
                i.update(rate)
        if i.item:
            price = frappe.db.sql("""select price_list_rate from `tabItem Price` where item_code=%s order by valid_from desc limit 1""",i.item,as_dict=1)
            if price:
                p = {"price_list_rate":price[0].price_list_rate}
                i.update(p)
            i_name = frappe.db.get_value("Item",i.item,"item_name")
            item_name = {"item_name":i_name}
            i.update(item_name)
            s_uom = frappe.db.get_value("Item",i.item,"stock_uom")
            uom = {"uom":s_uom}
            i.update(uom)
            desc = frappe.db.get_value("Item",i.item,"description")
            desc = {"description":desc}
            i.update(desc)
            i_acc = frappe.db.sql("""select income_account from `tabItem Default` where parent=%s and company=%s""",(i.item,company),as_dict=1 )
            if i_acc:
                income_account = {"income_account":i_acc[0].income_account}
                i.update(income_account)
           
    return datas

    
@frappe.whitelist()
def get_item_details(name,company):
    data = json.loads(name)
    items = []

    for i in data:
        items += frappe.db.sql(""" SELECT name,item,total_working_hour,total_costing_rate_before_deduction,total_costing_hour,reference_type,employee_code,employee_name,staff_code,staff_name FROM `tabTimesy` WHERE name=%s""", i, as_dict=1)

    for i in items:
        s_iq_id=''
        s_nat = ''
        iq_id=''
        nat = ''
        if i.employee_code:    
            iq_id = frappe.db.get_value("Employee",i.employee_code,"iqama_id") 
            nation = frappe.db.get_value("Employee",i.employee_code,"nationality") 
            i_id ={"iq_id":iq_id}
            i.update(i_id)
            nat ={"nation":nation}
            i.update(nat)
            price = frappe.db.sql("""select default_cost_rate_per_hour from `tabStaffing Cost` where employee_code=%s and status='Active'""",i.employee_code,as_dict=1)
            if price:
                rate = {"hourly_rate":price[0].default_cost_rate_per_hour}
                i.update(rate)
            else:
                rate = {"hourly_rate":0}
                i.update(rate)
        if i.staff_code:
            
            s_iq_id = frappe.db.get_value("Staff",i.staff_code,"iqama_number") 
            s_nation = frappe.db.get_value("Staff",i.staff_code,"nationality") 
            s_i_id ={"s_iq_id":s_iq_id}
            i.update(s_i_id)
            s_nat ={"s_nation":s_nation}
            i.update(s_nat)
            price = frappe.db.sql("""select default_cost_rate_per_hour from `tabStaffing Cost` where staff_code=%s and status='Active'""",i.staff_code,as_dict=1)
            if price:
                rate = {"hourly_rate":price[0].default_cost_rate_per_hour}
                i.update(rate)
            else:
                rate = {"hourly_rate":0}
                i.update(rate)
        if i.item:
            i_name = frappe.db.get_value("Item",i.item,"item_name")
            item_name = {"item_name":i_name}
            i.update(item_name)
            s_uom = frappe.db.get_value("Item",i.item,"stock_uom")
            uom = {"uom":s_uom}
            i.update(uom)
            desc = frappe.db.get_value("Item",i.item,"description")
            desc = {"description":desc}
            i.update(desc)
            i_acc = frappe.db.sql("""select income_account from `tabItem Default` where parent=%s and company=%s""",(i.item,company),as_dict=1 )
            if i_acc:
                income_account = {"income_account":i_acc[0].income_account}
                i.update(income_account)

            

    return items


@frappe.whitelist()
def get_staffing(doctype, target,setters,d,e,filters):
    print("============================")
    print(doctype)
    print(target)
    print(setters)
    print(d)
    print(e)
    print(filters)
    data = []
    condition = ""
    if target:
        condition += " and (name like '%{0}%') ".format(target)

    if 'customer_name' in filters:
        condition += " and customer_name like '{0}' ".format(filters['customer_name'][1])

    if "staffing_type" in filters:
        condition += get_condition(filters)

    if "start_date" in filters:
        condition += " and start_date = '{0}' ".format(filters['start_date'])

    if "supplier_name" in filters:
        condition += " and supplier_name like '{0}' ".format(filters['supplier_name'][1])

    if "employee_name" in filters:
        condition += " and (employee_name like '{0}' or staff_name like '{1}') ".format(filters['employee_name'][1],filters['employee_name'][1])

    time_list = ""
    if "doctype" in filters:
        time_list += " and parenttype = '{0}'".format(filters['doctype'])

    print(condition)
    query = """ SELECT * FROM `tabTimesy` WHERE docstatus=1 and status='Completed' {0}""".format(condition)
    print(query)
    staffing = frappe.db.sql(query, as_dict=1)
    for i in staffing:
        check = frappe.db.sql(""" SELECT * FROm `tabTimesy List` WHERE timesy=%s {0}""".format(time_list), i.name, as_dict=1)
        if len(check) == 0:
            data.append({
                "name": i.name,
                "staffing_type": i.staffing_type,
                "employee_name": i.staff_name if i.reference_type == 'Staff' else i.employee_name,
                "customer_name": i.customer_name,
                "supplier_name": i.supplier_name,
                "start_date": i.start_date
            })
    print(data)
    return data


@frappe.whitelist()
def get_bulk_timesy(name):
    data = json.loads(name)
    timesies = []

    for j in data:
        timesies += frappe.db.sql(""" SELECT * FROM `tabTimesy` WHERE name=%s""", j, as_dict=1)
        if timesies:
            for i in timesies:
                if i.item:
                    item = frappe.db.sql("""select * from `tabItem` where name=%s""",i.item,as_dict=1)
                    i_name = {'item_name' : item[0].item_name}
                    i_uom = {'stock_uom':item[0].stock_uom}
                    i.update(i_name)
                    i.update(i_uom)

                if i.reference_type=="Staff":
                    staff_data  = frappe.db.sql("""select * from `tabStaff` where name=%s""",i.staff_code,as_dict=1)
                    s_id = {'staff_iqama_id' : staff_data[0].iqama_number}
                    s_nation = {'staff_nationality':staff_data[0].nationality}
                    i.update(s_id)
                    i.update(s_nation)
                    rate = frappe.db.get_value("Staffing Cost",{"staff_code":i.staff_code},"default_billing_rate_per_hour")
                    r = {"hourly_rate":rate}
                    i.update(r)

                if i.reference_type=="Employee":
                    employee_data  = frappe.db.sql("""select * from `tabEmployee` where name=%s""",i.employee_code,as_dict=1)
                    e_id = {'iqama_id' : employee_data[0].iqama_id}
                    e_nation = {'nationality':employee_data[0].nationality}
                    i.update(e_id)
                    i.update(e_nation)


    return timesies